bob/src/hw_metrics_collector.rs

```diff
@@ -80,32 +83,41 @@ impl HWMetricsCollector {
         }
     }
 
+    fn to_cpath(path: &Path) -> Vec<u8> {
+        use std::{ffi::OsStr, os::unix::ffi::OsStrExt};
+    
+        let path_os: &OsStr = path.as_ref();
+        let mut cpath = path_os.as_bytes().to_vec();
+        cpath.push(0);
+        cpath
+    }
+
     // FIXME: maybe it's better to cache needed disks, but I am not sure, that they would be
     // refreshed, if I clone them
     // NOTE: HashMap contains only needed mount points of used disks, so it won't be really big,
     // but maybe it's more efficient to store disks (instead of mount_points) and update them one by one
-    fn space(sys: &System, disks: &HashMap<PathBuf, String>) -> (u64, u64) {
-        sys.disks()
-            .iter()
-            .filter_map(|disk| {
-                disks
-                    .get(disk.mount_point())
-                    .map(|diskname| (disk, diskname))
-            })
-            .fold((0, 0), |(total, free), (disk, diskname)| {
-                let disk_total = bytes_to_mb(disk.total_space());
-                let disk_free = bytes_to_mb(disk.available_space());
-                trace!(
-                    "{} (with path {}): total = {}, free = {};",
-                    diskname,
-                    disk.mount_point()
-                        .to_str()
-                        .expect("Can't parse mount point"),
-                    disk_total,
-                    disk_free
-                );
-                (total + disk_total, free + disk_free)
-            })
+    fn space(disks: &HashMap<PathBuf, String>) -> (u64, u64, u64) {
+        let mut total = 0;
+        let mut used = 0;
+        let mut free = 0;
+        
+        for mount_point in disks.keys() {
+            let cm_p = Self::to_cpath(mount_point.as_path());
+            unsafe {
+                let mut stat: statvfs = std::mem::zeroed();
+                if statvfs(cm_p.as_ptr() as *const _, &mut stat) == 0 {
+                    let bsize = Wrapping(stat.f_bsize as u64);
+                    let blocks = Wrapping(stat.f_blocks as u64);
+                    let bavail = Wrapping(stat.f_bavail as u64);
+                    let bfree = Wrapping(stat.f_bfree as u64);
+                    total += (bsize * blocks).0;
+                    free += (bsize * bavail).0;
+                    used += ((blocks - bfree) * bsize).0;
+                }
+            }
+        }
+
+        (total, free, used)
```

Comment:
It's better to use proper struct here, as all members are of the same type, and it's easy to misuse result